MLflow 3.0 docs (#15078)
mlflow/mlflow: docs/docs/mlflow-3/genai-agent.mdx

---
sidebar_position: 2
---

import { APILink } from "@site/src/components/APILink";

# GenAI Agent with MLflow 3.0

This example demonstrates how to use MLflow to trace and evaluate OpenAI requests with prompts engineering.
It showcases how to register prompts, generate traces, and assess response performance using evaluation datasets.
The example also highlights the ability to track interactive traces and link them to the logged model for better observability.

> Prerequisites: run ``pip install --upgrade mlflow>=3.0.0rc0 --pre`` ``pip install openai`` to install MLflow 3.0 and OpenAI packages.

Reviewer: ````suggestion
Prerequisites:

Run the following command to install MLflow 3.0 and OpenAI packages.

```
pip install --upgrade mlflow>=3.0.0rc0 --pre
pip install openai
```
````

nit

    eval_dataset = mlflow.data.from_pandas(
        df=eval_df,
        name="eval_dataset",
        targets="expected_response",
        predictions="predictions",
    )
    mlflow.log_input(
        dataset=eval_dataset, model=LoggedModelInput(logged_model.model_id)
    )
    # Run the evaluation based on extra metrics
    result = mlflow.evaluate(
        data=eval_dataset,
        model_type="question-answering",
        extra_metrics=[
            mlflow.metrics.latency(),
            mlflow.metrics.genai.answer_correctness("openai:/gpt-4o"),
        ],
        # This is needed since answer_correctness looks for 'inputs' field
        evaluator_config={"col_mapping": {"inputs": "messages"}},
    )

print(result.tables["eval_results_table"])
#                                         messages  ...                answer_correctness/v1/justification
# 0  What is MLflow Tracking and how does it work?  ...  The output is mostly correct, providing accura...
# 1                         What is Unity Catalog?  ...  The output is completely incorrect as it descr...
# 2        What are user-defined functions (UDFs)?  ...  The output provides a general definition of us...
```

> Navigating to the evaluation run, you can see the metrics and their details displayed in the MLflow UI. This includes metrics like latency and answer correctness, providing insights into the model's performance.

Reviewer 2: Can we also show a screenshot of the model page with the new evaluation result?

Prerequisite: ``pip install langchain langchain-community langchain-openai``

## Create an agent

First, we create an agent and utilize MLflow's prompt registry capability to manage and link prompts effectively.

```python
from langchain_core.prompts import ChatPromptTemplate
from langchain_openai import ChatOpenAI

import mlflow

# register a prompt so we can link it when logging the model
system_prompt = mlflow.register_prompt(
    name="system_prompt",
    template="You are a chatbot that can answer questions about IT.",
    commit_message="Initial version of system prompt",
)

chat_model = ChatOpenAI(
    model="gpt-4o-mini", temperature=0.1, max_completion_tokens=2000
)
chat_prompt = ChatPromptTemplate.from_messages(
    [("system", system_prompt.to_single_brace_format()), ("user", "{messages}")]
)
chain = chat_prompt | chat_model
print(chain.invoke({"messages": "What is MLflow?"}))
```

Switch to the **Prompts** tab to view the registered prompt and its details:

<div className="center-div" style={{ width: "100%" }}>
    ![The MLflow UI showing a prompt version](/images/mlflow-3/genai/system_prompt.png)
</div>

## Log the agent

> We recommend using a code-based logging approach for agents. Save your agent to a python file (e.g., `agent.py`)

Reviewer 2: @serena-ruan Can we stop recommending code based logging and instead show `mlflow.create_external_model()` instead? This should be a small change.

        lc_model="agent.py",
        # specify the name of the model
        name="chain",
        params={"temperature": 0.1, "max_tokens": 2000},
        model_type="agent",
        input_example={"messages": "What is MLflow?"},
        # specify the prompts used, it will be linked to the run
        prompts=[system_prompt],
    )

# Inspect the LoggedModel and its properties
logged_model = mlflow.get_logged_model(model_info.model_id)
print(logged_model.model_id, logged_model.params)
# c981a134-0e42-4765-9e95-2d10f77023a8 {'max_tokens': '2000', 'temperature': '0.1'}
```

Check out the **Models** tab in the experiment to view the newly logged model named `chain` and its parameters.

<div className="center-div" style={{ width: "100%" }}>
    ![The MLflow UI showing the logged models in an experiment](/images/mlflow-3/genai/logged_models_tab.png)
</div>

On the **Logged Model** page, you can view detailed information, including the `model_id`, the source run of the model, and its parameters.

<div className="center-div" style={{ width: "100%" }}>
    ![The MLflow UI showing the logged model details page](/images/mlflow-3/genai/logged_model_page.png)
</div>

> ## Test the agent with tracing observability

Reviewer 2: ```suggestion
## Test the agent with tracing observability

In this section, we manual test the agent with some example queries and use MLflow Tracing to analyze the outputs and debug any issues.
```

#                          request_id                                              trace   timestamp_ms  ...                                              spans                                               tags assessments
# 0  c4922d6515004439ba6745506969c401  Trace(request_id=c4922d6515004439by6745506969c...  1742786913864  ...  [{'name': 'RunnableSequence', 'context': {'spa...  {'mlflow.artifactLocation': 'file:///Users/ser...          []
# 1  a6e4de782b5540f1af2f2661ba85abd2  Trace(request_id=a6e4de782b5540f1af2f2661by85a...  1742786909657  ...  [{'name': 'RunnableSequence', 'context': {'spa...  {'mlflow.artifactLocation': 'file:///Users/ser...          []
# 2  bc72367e96ae4e978e28594d5ce438e8  Trace(request_id=bc72367e96ae4e978e28594d5ce43...  1742786905377  ...  [{'name': 'RunnableSequence', 'context': {'spa...  {'mlflow.artifactLocation': 'file:///Users/ser...          []
```

Navigating to the **Traces** tab of the model's page, you can view the traces that were just generated.

<div className="center-div" style={{ width: "100%" }}>
    ![The MLflow UI showing the logged model autolog traces lineage](/images/mlflow-3/genai/logged_model_autolog_traces.png)
</div>

> ## Evaluate on the agent

Reviewer 2: ```suggestion
## Evaluate the agent's performance
```

> Then we use MLflow evaluate to assess the performance of the logged model on the evaluation dataset. 

Reviewer 2: ```suggestion
Finally, we use `mlflow.evaluate()` to assess the performance of the logged model on an evaluation dataset. 
```

Reviewer 2: Can we make this an  <APILink>? (I added backticks instead just to communicate the desired phrasing)
